### PR TITLE
A few descriptor improvements to prepare for Taproot support

### PR DESCRIPTION
**Pull request by sipa**
These are a few refactors and non-invasive improvements to the descriptors code to prepare for adding Taproot descriptors.

None of the commits change behavior in any way, except the last one which improves error reporting a bit.

### CODE REVIEW

**Review by RiccardoMasutti — APPROVED** [review]
ACK, seems good to me. Waiting for squash and will review again.

**Comment by sipa** [comment]
I don't intend to squash. They're all independent improvements.

File: src/script/descriptor.cpp [hunk 1]
```diff
@@ -576,33 +581,37 @@ class DescriptorImpl : public Descriptor
...
         std::vector<CScript> subscripts;
-        if (m_subdescriptor_arg) {
-            FlatSigningProvider subprovider;
-            if (!m_subdescriptor_arg->ExpandHelper(pos, arg, read_cache, subscripts, subprovider, write_cache)) return false;
-            out = Merge(out, subprovider);
+        FlatSigningProvider subprovider;
+        for (const auto& subarg : m_subdescriptor_args) {
+            std::vector<CScript> outscripts;
+            if (!subarg->ExpandHelper(pos, arg, read_cache, outscripts, subprovider, write_cache)) return false;
+            if (outscripts.size() != 1 && m_subdescriptor_args.size() != 1) return false;
```
Comment (Sjors) [inline src/script/descriptor.cpp:600]: 5a044ac64f3f51dc89a772f512d403c7baa12ae2: I'm a bit confused what this new check is doing

File: src/script/descriptor.cpp [hunk 2]
```diff
@@ -576,33 +581,37 @@ class DescriptorImpl : public Descriptor
...
+        out = Merge(std::move(out), std::move(subprovider));
 
         std::vector<CPubKey> pubkeys;
         pubkeys.reserve(entries.size());
         for (auto& entry : entries) {
             pubkeys.push_back(entry.first);
             out.origins.emplace(entry.first.GetID(), std::make_pair<CPubKey, KeyOriginInfo>(CPubKey(entry.first), std::move(entry.second)));
         }
-        if (m_subdescriptor_arg) {
+        if (m_subdescriptor_args.size() == 1) {
```
Comment (Sjors) [inline src/script/descriptor.cpp:611]: 5a044ac64f3f51dc89a772f512d403c7baa12ae2: why is `m_subdescriptor_args` length 1 a special case?

**Review by Sjors — COMMENTED** [review]
Code review 45bc0c04e0f2ea6dcce734652f45cabc6d10d7d5 looks good, but I got confused in one place...

**Comment by achow101** [comment]
Partial ACK. Same questions as Sjors.

**Comment by sipa** [comment]
@achow101 @Sjors Good that you raise that point.

The original reason why this behavior of MakeScripts-is-invoked-once-per-evaluation existed was to support things like `combo()`, where a single evaluation at a single index results in multiple scriptPubKeys. This was particularly complicated to combine with descriptors that support multiple subdescriptors (like `tr()`) - there isn't really a straightforward way of doing that (cartesian product? requiring the same number of evaluations from each and combining pairwise? ...?). I didn't want to deal with that, so I added the restriction that this once-per-evaluation only existed for descriptors with 1 subdescriptor.

This is all silly, because we don't ever need it. `combo()` is the only descriptor that produces multiple output scripts, and it is only permitted at the top level - so it's not even possible to have it as a subdescriptor. I've just removed support for multiple-outputscripts-from-subscriptor entirely in a separate commit now. Bye.

File: src/script/descriptor.cpp (hunk 2 above)
Comment (sipa) [reply src/script/descriptor.cpp]: Gone.

File: src/script/descriptor.cpp (hunk 1 above)
Comment (sipa) [reply src/script/descriptor.cpp]: Gone.

**Comment by Sjors** [comment]
Thanks, I'll review again after rebase.

**Comment by sipa** [comment]
Rebased.

**Comment by Sjors** [comment]
ACK 05c633e

Conceptually I find it marginally simpler if you reverse the first and second commit, so the first step is to remove the unused loop:

```cpp
        if (m_subdescriptor_arg) {
            const auto& subscript = subscripts[0];
            out.scripts.emplace(CScriptID(subscript), subscript);
            output_scripts = MakeScripts(pubkeys, &subscript, out);
        } else {
            output_scripts = MakeScripts(pubkeys, nullptr, out);
        }
 ```
 
 And then the second commit drops `out.scripts.emplace` and moves that into `MakeScripts` (which is also where `combo` does this).
 
To sanity check my understanding: `key_exp_index` is currently used for multisig descriptors to create a seperate cache for each participating key. With taproot, you might have a root key (`key_exp_index == 0`), one branch with a key ``key_exp_index == 2`, and another branch with a 3 key multisig (`key_exp_index  3,4,5). Is that how it works?

**Comment by achow101** [comment]
ACK 05c633e0efaf300aa3353303f77c0a622647a588

**Comment by sipa** [comment]
@Sjors 

> Conceptually I find it marginally simpler if you reverse the first and second commit

Ok, done.

> With taproot, you might have a root key (`key_exp_index == 0`), one branch with a key `key_exp_index == 2`, and another branch with a 3 key multisig (`key_exp_index  3,4,5`). Is that how it works?

I think you skipped the 1, but otherwise: exactly.

**Comment by achow101** [comment]
re-ACK 35cbef1c483ac9905331f8035d32600e105baab4

Only change was commit reordering.

File: src/script/descriptor.cpp [hunk 3]
```diff
@@ -480,34 +480,36 @@ class DescriptorImpl : public Descriptor
...
-    virtual std::vector<CScript> MakeScripts(const std::vector<CPubKey>& pubkeys, const CScript* script, FlatSigningProvider& out) const = 0;
+    virtual std::vector<CScript> MakeScripts(const std::vector<CPubKey>& pubkeys, Span<const CScript> scripts, FlatSigningProvider& out) const = 0;
 
 public:
-    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, std::unique_ptr<DescriptorImpl> script, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_arg(std::move(script)) {}
+    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_args() {}
+    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, std::unique_ptr<DescriptorImpl> script, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_args(Vector(std::move(script))) {}
+    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, std::vector<std::unique_ptr<DescriptorImpl>> scripts, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_args(std::move(scripts)) {}
```
Comment (Sjors) [inline src/script/descriptor.cpp:507]: warning: unused member function 'DescriptorImpl'

**Review by Sjors — COMMENTED** [review]
re-ACK 35cbef1c483ac9905331f8035d32600e105baab4 modulo unused constructor

**Review by S3RK — COMMENTED** [review]
Code Review ACK 394e387. Nice and clear improvements!

File: src/script/descriptor.cpp [hunk 4]
```diff
@@ -577,30 +579,29 @@ class DescriptorImpl : public Descriptor
...
-            assert(subscripts.size() == 1);
-            out = Merge(out, subprovider);
+        FlatSigningProvider subprovider;
+        for (const auto& subarg : m_subdescriptor_args) {
+            std::vector<CScript> outscripts;
+            if (!subarg->ExpandHelper(pos, arg, read_cache, outscripts, subprovider, write_cache)) return false;
+            assert(outscripts.size() == 1);
+            subscripts.emplace_back(std::move(outscripts[0]));
         }
+        out = Merge(std::move(out), std::move(subprovider));
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/script/descriptor.cpp:601]: It doesn't matter now apparently, as Merge takes const lvalue references only. But a more efficient variant could be added that takes advantage of inputs that aren't needed anymore.

File: src/script/descriptor.cpp [hunk 5]
```diff
@@ -543,12 +554,7 @@ class DescriptorImpl : public Descriptor
             }
             ret += std::move(tmp);
         }
-        for (const auto& scriptarg : m_subdescriptor_args) {
-            if (pos++) ret += ",";
-            std::string tmp;
-            if (!scriptarg->ToStringHelper(arg, tmp, priv, normalized)) return false;
-            ret += std::move(tmp);
-        }
+        if (!ToStringSubScriptHelper(arg, ret, priv, normalized, pos)) return false;
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/script/descriptor.cpp:557]: Done.

File: src/script/descriptor.cpp [hunk 6]
```diff
@@ -621,9 +622,9 @@ class DescriptorImpl : public Descriptor
             if (!p->GetPrivKey(pos, provider, key)) continue;
             out.keys.emplace(key.GetPubKey().GetID(), key);
         }
-        if (m_subdescriptor_arg) {
+        for (const auto& arg : m_subdescriptor_args) {
             FlatSigningProvider subprovider;
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/script/descriptor.cpp]: Done. In fact, I don't need any at all.

File: src/script/descriptor.cpp (hunk 3 above)
Comment (sipa) [reply src/script/descriptor.cpp]: Fixed.

**Comment by sipa** [comment]
Rebased, and addressed a few comments.

**Comment by S3RK** [comment]
reACK 0b188b7
The only changes are addressing review comments above. 1) removing unused constructor 2) removing redundant signing provided instance creation 3) simplifying `ToStringSubScriptHelper` signature

**Comment by Sjors** [comment]
re-ACK 0b188b7

**Comment by achow101** [comment]
re-ACK 0b188b751f970027c52729e0c223cc9257669322

**Comment by Sjors** [comment]
@instagibbs what do you think?

**Comment by instagibbs** [comment]
Seems like 3 ACKs is enough? not familiar enough with this code anymore to give a proper ACK

File: src/script/descriptor.cpp [hunk 7]
```diff
@@ -1045,6 +1066,9 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t key_exp_index, Span<const c
         }
         auto bytes = ParseHex(str);
         return std::make_unique<RawDescriptor>(CScript(bytes.begin(), bytes.end()));
+    } else if (Func("raw", expr)) {
+        error = "Can only have raw() at top level";
```
Comment (ysangkok) [inline src/script/descriptor.cpp:1070]: why is there no new test for this and for `addr()`?

**Pull request merged**